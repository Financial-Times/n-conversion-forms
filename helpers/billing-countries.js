import { billingCountries } from 'n-common-static-data';

export { billingCountries };
