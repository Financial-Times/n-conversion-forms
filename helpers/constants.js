const printRegions = {
	cemeaV1: 'CEMEA_V1',
	cemeaV2: 'CEMEA_V2',
	apac: 'APAC',
};

const addressLine1AutoComplete = 'address-line1';
const addressLine2AutoComplete = 'address-line2';
const addressLine3AutoComplete = 'address-line3';

module.exports = {
	printRegions,
	addressLine1AutoComplete,
	addressLine2AutoComplete,
	addressLine3AutoComplete,
};
