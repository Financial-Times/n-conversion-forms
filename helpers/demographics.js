const { demographics } = require('n-common-static-data');

export { demographics };
