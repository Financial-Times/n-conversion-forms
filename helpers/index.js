module.exports = {
	'ncf-countries': require('./ncf-countries'),
	'ncf-common-data': require('./ncf-common-data'),
	cemeaV1ISO: require('./supportedCountries').cemeaV1ISO,
	cemeaV2ISO: require('./supportedCountries').cemeaV2ISO,
	apacISO: require('./supportedCountries').apacISO,
	identifyFTShippingZone: require('./supportedCountries')
		.identifyFTShippingZone,
	countriesSupported: require('./supportedCountries').countriesSupported,
	countriesSupportedISO: require('./supportedCountries').countriesSupportedISO,
	supportedPostcodeValidators: require('./supportedPostcodeValidators')
		.supportedPostcodeValidators,
	allSupportedPostcodeExamples: require('./supportedPostcodeExamples')
		.allSupportedPostcodeExamples,
	demographics: require('./demographics').demographics,
	billingCountries: require('./billing-countries').billingCountries,
};
