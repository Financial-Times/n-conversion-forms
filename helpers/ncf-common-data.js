const _get = require('lodash.get');
const COMMON_DATA = require('n-common-static-data');

module.exports = function ({ hash, fn }) {
	const context = Object.assign({}, { [hash.export]: _get(COMMON_DATA, hash.import) }, this);
	return fn(context);
};
