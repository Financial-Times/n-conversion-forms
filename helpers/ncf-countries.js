const { countries } = require('n-common-static-data').billingCountries;

const MINIMUM_TO_GROUP = 50;
const MINIMUM_TO_SHOW_GROUPS = 2;

module.exports = function ({ hash = {}, fn }) {
	const filter = Array.isArray(hash.filterList) ? hash.filterList : false;
	const value = hash.value;

	let data = countries;

	// Only show countries in the filter
	if (filter) {
		data = data.filter(item => filter.includes(item.code));
	}

	// Select the country
	if (value) {
		data = data.map(item => {
			item.selected = item.code === value;
			return item;
		});
	}

	// Group countries
	if (data.length >= MINIMUM_TO_GROUP) {
		data = groupCountries(data);
	}

	const context = Object.assign({ countries: data }, this);
	return fn(context);
};

/**
 * Produce a frequently used option group
 * @param {Array} countries
 * @return {Array}
 */
function groupCountries (countries) {
	const frequentlyUsed = ['GBR', 'USA', 'JPN', 'FRA', 'CAN'];

	const frequentlyUsedCountries = countries
		.filter(item => frequentlyUsed.includes(item.code))
		.map(item => Object.assign({}, item))
		.sort((a, b) => frequentlyUsed.indexOf(a.code) - frequentlyUsed.indexOf(b.code));

	// Not worth showing return standard countries list
	if (frequentlyUsedCountries.length < MINIMUM_TO_SHOW_GROUPS) {
		return countries;
	}

	// If selected country is in frequently used only leave that one selected
	// otherwise the user is shown the lowest selected option
	const selected = frequentlyUsedCountries.find(item => item.selected);
	if (selected) {
		countries.forEach(item => item.selected = false);
	}

	return [
		{ label: 'Frequently Used', countries: frequentlyUsedCountries },
		{ label: 'Alphabetical', countries }
	];
}
