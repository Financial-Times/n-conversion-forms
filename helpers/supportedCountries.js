const supportedCountriesMasterList = {
	cemeaV1: {
		AUT: 'Austria',
		BEL: 'Belgium',
		BGR: 'Bulgaria',
		HRV: 'Croatia',
		CZE: 'Czechia',
		DNK: 'Denmark',
		FIN: 'Finland',
		DEU: 'Germany',
		GRC: 'Greece',
		HUN: 'Hungary',
		ISL: 'Iceland',
		ITA: 'Italy',
		LIE: 'Liechtenstein',
		LTU: 'Lithuania',
		NLD: 'Netherlands',
		NOR: 'Norway',
		POL: 'Poland',
		PRT: 'Portugal',
		QAT: 'Qatar',
		ROU: 'Romania',
		RUS: 'Russian Federation',
		SVK: 'Slovakia',
		SVN: 'Slovenia',
		ZAF: 'South Africa',
		ESP: 'Spain',
		SWE: 'Sweden',
		CHE: 'Switzerland'
	},
	cemeaV2: { FRA: 'France', LUX: 'Luxembourg', MCO: 'Monaco' },
	apac: {
		CHN: 'China',
		HKG: 'Hong Kong',
		IND: 'India',
		IDN: 'Indonesia',
		JPN: 'Japan',
		KOR: 'Korea, Republic of',
		MYS: 'Malaysia',
		PHL: 'Philippines',
		SGP: 'Singapore',
		TWN: 'Taiwan, Province of China',
		THA: 'Thailand'
	},
	GBR: 'United Kingdom',
	USA: 'United States',
	CAN: 'Canada'
};


const cemeaV1ISO = Object.keys(supportedCountriesMasterList.cemeaV1);
const cemeaV2ISO = Object.keys(supportedCountriesMasterList.cemeaV2);
const apacISO = Object.keys(supportedCountriesMasterList.apac);
const countriesSupported = Object.assign({}, ...(function _flatten (o) { return [].concat(...Object.keys(o).map(k => typeof o[k] === 'object' ? _flatten(o[k]) : ({[k]: o[k]})));}(supportedCountriesMasterList)));


module.exports = { cemeaV1ISO, cemeaV2ISO, apacISO, countriesSupported };
