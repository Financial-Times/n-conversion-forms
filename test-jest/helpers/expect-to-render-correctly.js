import ReactDOMServer from 'react-dom/server';

export const expectToRenderCorrectly = {
	toRenderCorrectly(reactComponent, props) {
		// Render the React component with props.
		const ReactComponentRendered = ReactDOMServer.renderToStaticMarkup(
			reactComponent(props)
		);

		// eslint-disable-next-line jest/no-standalone-expect
		expect(ReactComponentRendered).toMatchSnapshot();

		// This point is reached when the above assertions were successful.
		// The test should therefore always pass, that means it needs to be
		// `true` when used normally, and `false` when `.not` was used.
		return {
			pass: !this.isNot,
		};
	},
};
