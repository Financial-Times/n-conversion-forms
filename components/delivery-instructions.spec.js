import { DeliveryInstructions } from './index';
import { expectToRenderCorrectly } from '../test-jest/helpers/expect-to-render-correctly';

import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });

const spanMessageByCountry = {
	GBR: 'For newspaper delivery, we can only deliver to the ground floor, so if you live in an apartment, we’ll leave the newspaper at reception or by the entrance. We deliver in the early hours of the morning so our drivers won’t be able to contact you or ring your doorbell.',
	USA: 'Please note we cannot guarantee delivery of the newspaper to a specific location on your property, which also includes delivery to a specific floor/suite in a building. US Federal Law prohibits delivery of newspapers into a mailbox, except via a USPS mail carrier. If you prefer delivery to a mailbox, please either select the "PO Box" delivery option from the top of this form, or contact. FT Customer Care',
	CAN: 'Please note we cannot guarantee delivery of the newspaper to a specific location on your property, which also includes delivery to a specific floor/suite in a building. If you prefer delivery by Canada Post, please either select the "PO Box" delivery option from the top of this form, or contact. FT Customer Care',
};

expect.extend(expectToRenderCorrectly);

describe('DeliveryInstructions', () => {
	it('renders with default props', () => {
		const props = { placeholder: 'test', country: 'GBR' };

		expect(DeliveryInstructions).toRenderCorrectly(props);
	});

	it('renders with an error', () => {
		const props = { placeholder: 'test', hasError: true, country: 'GBR' };

		expect(DeliveryInstructions).toRenderCorrectly(props);
	});

	it('renders with maxlength', () => {
		const props = { placeholder: 'test', maxlength: 200, country: 'GBR' };

		expect(DeliveryInstructions).toRenderCorrectly(props);
	});

	it('renders with rows', () => {
		const props = { placeholder: 'test', rows: 20, country: 'GBR' };

		expect(DeliveryInstructions).toRenderCorrectly(props);
	});

	it('renders with a custom value', () => {
		const props = { placeholder: 'test', value: 'foobar', country: 'GBR' };

		expect(DeliveryInstructions).toRenderCorrectly(props);
	});

	it('renders with a disabled input element', () => {
		const props = { placeholder: 'test', isDisabled: true, country: 'GBR' };

		expect(DeliveryInstructions).toRenderCorrectly(props);
	});

	it('renders with a USA span message', () => {
		const props = { placeholder: 'test', country: 'USA' };

		expect(DeliveryInstructions).toRenderCorrectly(props);
	});

	it('renders with a CAN span message', () => {
		const props = { placeholder: 'test', country: 'CAN' };

		expect(DeliveryInstructions).toRenderCorrectly(props);
	});

	it('renders as with a GBR span message and check the text', () => {
		const props = { isRequired: false, country: 'GBR' };

		const component = mount(DeliveryInstructions(props));

		const element = component.find('.o-forms-title__prompt').first();
		expect(element.text()).toBe(spanMessageByCountry['GBR']);
	});

	it('renders as with a USA span message and check the text', () => {
		const props = { isRequired: false, country: 'USA' };

		const component = mount(DeliveryInstructions(props));

		const element = component.find('.o-forms-title__prompt').first();
		expect(element.text()).toBe(spanMessageByCountry[props.country]);
	});

	it('renders as with a CAN span message and check the text', () => {
		const props = { isRequired: false, country: 'CAN' };

		const component = mount(DeliveryInstructions(props));

		const element = component.find('.o-forms-title__prompt').first();
		expect(element.text()).toBe(spanMessageByCountry[props.country]);
	});
});
