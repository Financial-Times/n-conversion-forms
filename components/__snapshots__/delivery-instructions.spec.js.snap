// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeliveryInstructions renders with a CAN span message 1`] = `
<label id="deliveryInstructionsField"
       class="o-forms-field o-forms-field--optional"
       data-validate="required"
       for="deliveryInstructions"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Delivery instructions
    </span>
    <span class="o-forms-title__prompt">
      Please note we cannot guarantee delivery of the newspaper to a specific location on your property, which also includes delivery to a specific floor/suite in a building. If you prefer delivery by Canada Post, please either select the &quot;PO Box&quot; delivery option from the top of this form, or contact.
      <a href="https://help.ft.com/contact/">
        FT Customer Care
      </a>
    </span>
  </span>
  <span class="o-forms-input o-forms-input--textarea">
    <textarea id="deliveryInstructions"
              name="deliveryInstructions"
              data-trackable="field-deliveryInstructions"
              placeholder="test"
    >
    </textarea>
  </span>
</label>
`;

exports[`DeliveryInstructions renders with a USA span message 1`] = `
<label id="deliveryInstructionsField"
       class="o-forms-field o-forms-field--optional"
       data-validate="required"
       for="deliveryInstructions"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Delivery instructions
    </span>
    <span class="o-forms-title__prompt">
      Please note we cannot guarantee delivery of the newspaper to a specific location on your property, which also includes delivery to a specific floor/suite in a building. US Federal Law prohibits delivery of newspapers into a mailbox, except via a USPS mail carrier. If you prefer delivery to a mailbox, please either select the &quot;PO Box&quot; delivery option from the top of this form, or contact.
      <a href="https://help.ft.com/contact/">
        FT Customer Care
      </a>
    </span>
  </span>
  <span class="o-forms-input o-forms-input--textarea">
    <textarea id="deliveryInstructions"
              name="deliveryInstructions"
              data-trackable="field-deliveryInstructions"
              placeholder="test"
    >
    </textarea>
  </span>
</label>
`;

exports[`DeliveryInstructions renders with a custom value 1`] = `
<label id="deliveryInstructionsField"
       class="o-forms-field o-forms-field--optional"
       data-validate="required"
       for="deliveryInstructions"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Delivery instructions
    </span>
    <span class="o-forms-title__prompt">
      For newspaper delivery, we can only deliver to the ground floor, so if you live in an apartment, we’ll leave the newspaper at reception or by the entrance. We deliver in the early hours of the morning so our drivers won’t be able to contact you or ring your doorbell.
    </span>
    <span class="o-forms-title__prompt">
      Special characters including punctuation cannot be used in the Delivery instructions.
    </span>
  </span>
  <span class="o-forms-input o-forms-input--textarea">
    <textarea id="deliveryInstructions"
              name="deliveryInstructions"
              data-trackable="field-deliveryInstructions"
              placeholder="test"
    >
      foobar
    </textarea>
  </span>
</label>
`;

exports[`DeliveryInstructions renders with a disabled input element 1`] = `
<label id="deliveryInstructionsField"
       class="o-forms-field o-forms-field--optional"
       data-validate="required"
       for="deliveryInstructions"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Delivery instructions
    </span>
    <span class="o-forms-title__prompt">
      For newspaper delivery, we can only deliver to the ground floor, so if you live in an apartment, we’ll leave the newspaper at reception or by the entrance. We deliver in the early hours of the morning so our drivers won’t be able to contact you or ring your doorbell.
    </span>
    <span class="o-forms-title__prompt">
      Special characters including punctuation cannot be used in the Delivery instructions.
    </span>
  </span>
  <span class="o-forms-input o-forms-input--textarea">
    <textarea id="deliveryInstructions"
              name="deliveryInstructions"
              data-trackable="field-deliveryInstructions"
              placeholder="test"
              disabled
    >
    </textarea>
  </span>
</label>
`;

exports[`DeliveryInstructions renders with an error 1`] = `
<label id="deliveryInstructionsField"
       class="o-forms-field o-forms-field--optional"
       data-validate="required"
       for="deliveryInstructions"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Delivery instructions
    </span>
    <span class="o-forms-title__prompt">
      For newspaper delivery, we can only deliver to the ground floor, so if you live in an apartment, we’ll leave the newspaper at reception or by the entrance. We deliver in the early hours of the morning so our drivers won’t be able to contact you or ring your doorbell.
    </span>
    <span class="o-forms-title__prompt">
      Special characters including punctuation cannot be used in the Delivery instructions.
    </span>
  </span>
  <span class="o-forms-input o-forms-input--textarea o-forms-input--invalid">
    <textarea id="deliveryInstructions"
              name="deliveryInstructions"
              data-trackable="field-deliveryInstructions"
              placeholder="test"
    >
    </textarea>
  </span>
</label>
`;

exports[`DeliveryInstructions renders with default props 1`] = `
<label id="deliveryInstructionsField"
       class="o-forms-field o-forms-field--optional"
       data-validate="required"
       for="deliveryInstructions"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Delivery instructions
    </span>
    <span class="o-forms-title__prompt">
      For newspaper delivery, we can only deliver to the ground floor, so if you live in an apartment, we’ll leave the newspaper at reception or by the entrance. We deliver in the early hours of the morning so our drivers won’t be able to contact you or ring your doorbell.
    </span>
    <span class="o-forms-title__prompt">
      Special characters including punctuation cannot be used in the Delivery instructions.
    </span>
  </span>
  <span class="o-forms-input o-forms-input--textarea">
    <textarea id="deliveryInstructions"
              name="deliveryInstructions"
              data-trackable="field-deliveryInstructions"
              placeholder="test"
    >
    </textarea>
  </span>
</label>
`;

exports[`DeliveryInstructions renders with maxlength 1`] = `
<label id="deliveryInstructionsField"
       class="o-forms-field o-forms-field--optional"
       data-validate="required"
       for="deliveryInstructions"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Delivery instructions
    </span>
    <span class="o-forms-title__prompt">
      For newspaper delivery, we can only deliver to the ground floor, so if you live in an apartment, we’ll leave the newspaper at reception or by the entrance. We deliver in the early hours of the morning so our drivers won’t be able to contact you or ring your doorbell.
    </span>
    <span class="o-forms-title__prompt">
      Special characters including punctuation cannot be used in the Delivery instructions.
    </span>
  </span>
  <span class="o-forms-input o-forms-input--textarea">
    <textarea id="deliveryInstructions"
              name="deliveryInstructions"
              maxlength="200"
              data-trackable="field-deliveryInstructions"
              placeholder="test"
    >
    </textarea>
  </span>
</label>
`;

exports[`DeliveryInstructions renders with rows 1`] = `
<label id="deliveryInstructionsField"
       class="o-forms-field o-forms-field--optional"
       data-validate="required"
       for="deliveryInstructions"
>
  <span class="o-forms-title">
    <span class="o-forms-title__main">
      Delivery instructions
    </span>
    <span class="o-forms-title__prompt">
      For newspaper delivery, we can only deliver to the ground floor, so if you live in an apartment, we’ll leave the newspaper at reception or by the entrance. We deliver in the early hours of the morning so our drivers won’t be able to contact you or ring your doorbell.
    </span>
    <span class="o-forms-title__prompt">
      Special characters including punctuation cannot be used in the Delivery instructions.
    </span>
  </span>
  <span class="o-forms-input o-forms-input--textarea">
    <textarea id="deliveryInstructions"
              name="deliveryInstructions"
              rows="20"
              data-trackable="field-deliveryInstructions"
              placeholder="test"
    >
    </textarea>
  </span>
</label>
`;
